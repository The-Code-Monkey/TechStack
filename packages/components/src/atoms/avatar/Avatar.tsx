import { IconTypes } from '@aw-web-design/react-feather';
import { get, TypographyProps } from '@aw-web-design/styled-system';
import { memo, useContext, useMemo } from 'react';
import { ThemeContext } from 'styled-components';

import { Box, Icon, BoxProps, Text } from '../../primal';

export interface Props extends BoxProps {
  iconName?: IconTypes;
  url?: string;
  children?: string;
  size?: string | number;
  typography?: TypographyProps;
}

export const Avatar = ({
  iconName,
  url,
  children,
  size = 10,
  typography,
  ...rest
}: Props) => {
  const { space } = useContext(ThemeContext);

  const memoSize = useMemo(() => get(space, size, size), [size, space]);

  return (
    <Box
      size={size}
      overflow='hidden'
      borderRadius={5}
      bg='neutrals.8'
      display='flex'
      justifyContent='center'
      alignItems='center'
      border='1'
      borderColor='common.black'
      {...rest}
    >
      {iconName && <Icon name={iconName} />}
      {url && <img src={url} width={memoSize} height={memoSize} />}
      {children && (
        <Text textAlign='center' {...typography}>
          {children
            .split(/\s/g)
            .map(s => s.substr(0, 1).toUpperCase())
            .join('')}
        </Text>
      )}
    </Box>
  );
};

export default memo(Avatar);
