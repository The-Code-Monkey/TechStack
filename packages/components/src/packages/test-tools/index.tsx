import { mount, shallow, ShallowWrapper, ReactWrapper } from 'enzyme';
import React, { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components';

import theme from '../theme';
import { ConfigContext } from '../utils';

const config = {
  outputDir: '/src/packages/',
  iconDir: 'feather',
};

const removeProperties = () => {
  const keys = ['theme', 'styledTheme'];
  return {
    test: (val: any) =>
      val &&
      typeof val === 'object' &&
      'props' in val &&
      Object.keys(val.props).some((prop) => keys.some((key) => key === prop)),
    print: (val: any, serialize: any) => {
      keys.forEach((key) => {
        delete val.props[key];
      });
      return serialize(val);
    },
  };
};

const formattedTheme = {
  ...theme,
  colors: {
    common: theme.colors?.common,
    modes: theme.colors?.modes,
    ...theme.colors?.modes.light,
  },
};

export const shallowWithTheme: (children: ReactNode) => ShallowWrapper = (
  children: ReactNode
) =>
  shallow(<ThemeProvider theme={formattedTheme}>{children}</ThemeProvider>)
    .dive()
    .shallow();

export const mountWithTheme: (children: ReactNode) => ReactWrapper = (
  children: ReactNode
) => {
  expect.addSnapshotSerializer(removeProperties());

  return mount(
    <ConfigContext.Provider value={config}>
      <ThemeProvider theme={formattedTheme}>
        <>{children}</>
      </ThemeProvider>
    </ConfigContext.Provider>
  );
};
