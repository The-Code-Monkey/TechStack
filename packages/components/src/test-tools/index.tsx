import isPropValid from '@emotion/is-prop-valid';
import { render, RenderResult } from '@testing-library/react';
import { ReactNode } from 'react';
import { StyleSheetManager, ThemeProvider } from 'styled-components';
import { expect } from 'bun:test';

import theme from '../theme';
import { ConfigContext } from '../utils';

const config = {
  outputDir: '/src/packages/',
  iconDir: 'feather',
};

const removeProperties = () => {
  const keys = ['theme', 'styledTheme'];
  return {
    test: (val: { props: unknown }) =>
      val &&
      typeof val === 'object' &&
      'props' in val &&
      Object.keys(val.props).some(prop => keys.some(key => key === prop)),
    serialize: (val, config, indentation, depth, refs, printer) => {
      keys.forEach(key => {
        delete val.props[key];
      });
      return printer(val, config, indentation, depth, refs);
    },
  };
};

const formattedTheme = {
  ...theme,
  colors: {
    common: theme.colors?.common,
    modes: theme.colors?.modes,
    ...theme.colors?.modes.light,
  },
};

export const mountWithTheme = (children: ReactNode): RenderResult => {
  // expect.addSnapshotSerializer(removeProperties());

  return render(
    <StyleSheetManager shouldForwardProp={isPropValid}>
      <ConfigContext.Provider value={config}>
        <ThemeProvider theme={formattedTheme}>
          <>{children}</>
        </ThemeProvider>
      </ConfigContext.Provider>
    </StyleSheetManager>
  );
};
