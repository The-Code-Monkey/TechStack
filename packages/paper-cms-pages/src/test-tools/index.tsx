import { ConfigContext } from '@techstack/components';
import { render } from '@testing-library/react';
import { ReactNode } from 'react';

import theme from '../theme';
import { ThemeProvider } from '../workarounds/styled-components';

const config = {
  outputDir: '/src/packages/',
  iconDir: 'feather',
};

const removeProperties = () => {
  const keys = ['theme', 'styledTheme'];
  return {
    test: (val: { props: unknown }) =>
      val &&
      typeof val === 'object' &&
      'props' in val &&
      Object.keys(val.props).some(prop => keys.some(key => key === prop)),
    serialize: (val, config, indentation, depth, refs, printer) => {
      keys.forEach(key => {
        delete val.props[key];
      });
      return printer(val, config, indentation, depth, refs);
    },
  };
};

const formattedTheme = {
  ...theme,
  colors: {
    common: theme.colors?.common,
    modes: theme.colors?.modes,
    ...theme.colors?.modes.light,
  },
};

export const mountWithTheme = (children: ReactNode) => {
  expect.addSnapshotSerializer(removeProperties());

  return render(
    <ConfigContext.Provider value={config}>
      <ThemeProvider theme={formattedTheme}>
        <>{children}</>
      </ThemeProvider>
    </ConfigContext.Provider>
  );
};
