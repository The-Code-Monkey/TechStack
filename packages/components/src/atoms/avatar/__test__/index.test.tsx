import { mountWithTheme } from '../../../test-tools';
import { Avatar } from '../Avatar';

describe('<Avatar />', () => {
  it('renders text correctly', () => {
    const { asFragment } = mountWithTheme(<Avatar>Test Text</Avatar>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders icon correctly', () => {
    const { asFragment } = mountWithTheme(<Avatar iconName='github' />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders img correctly', () => {
    const { asFragment } = mountWithTheme(
      <Avatar url='https://raw.githubusercontent.com/The-Code-Monkey/TechStack/feat/417-components-add-avatar/logo.png' />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
