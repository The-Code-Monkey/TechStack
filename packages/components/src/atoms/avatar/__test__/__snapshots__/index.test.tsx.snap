// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Avatar /> renders icon correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bscPHp hEFKjM"
    data-autoid="box"
  >
    <div
      class="sc-bscPHp gqgVuA sc-gqyZTE kmoqFI"
      data-autoid="github_icon"
    >
      <svg
        data-name="fallback-icon"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="10"
        />
        <path
          d="M9.09 9a3 3 0 0 1 5.83 1c0 2-3 3-3 3"
        />
        <line
          x1="12"
          x2="12.01"
          y1="17"
          y2="17"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Avatar /> renders img correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bscPHp hEFKjM"
    data-autoid="box"
  >
    <img
      height="64px"
      src="https://picsum.photos/id/237/200"
      width="64px"
    />
  </div>
</DocumentFragment>
`;

exports[`<Avatar /> renders text correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bscPHp hEFKjM"
    data-autoid="box"
  >
    <span
      class="sc-bKFsHG cMnIBd"
      data-autoid=""
    >
      TT
    </span>
  </div>
</DocumentFragment>
`;
